pkg/common/namespace_selection.go

```diff
@@ -103,3 +116,177 @@ func GetAllNamespaces(client kubernetes.Interface, labelSelector metav1.LabelSel
+	r.selections = make(map[string]NamespaceSelection)
+
+	neverEnqueue := predicate.NewPredicateFuncs(func(o client.Object) bool { return false })
+	mapToSingleton := func(_ client.Object) []reconcile.Request {
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Name: "NamespaceSelector",
+		}}}
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("NamespaceSelector").
+		For(
+			&policyv1.ConfigurationPolicy{},
+			builder.WithPredicates(neverEnqueue)).
+		Watches(
+			&source.Kind{Type: &corev1.Namespace{}},
+			handler.EnqueueRequestsFromMapFunc(mapToSingleton),
+			builder.WithPredicates(predicate.LabelChangedPredicate{})).
+		Complete(r)
+}
+
+type NamespaceSelection struct {
+	target     policyv1.Target
+	namespaces []string
+	hasUpdate  bool
+	err        error
+}
+
+func (r *NamespaceSelectorReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	log := logf.Log.WithValues("Reconciler", "NamespaceSelector")
+	oldSelections := make(map[string]NamespaceSelection)
+
+	r.lock.RLock()
+
+	for name, selection := range r.selections {
+		oldSelections[name] = selection
+	}
+
+	r.lock.RUnlock()
+
+	namespaces := corev1.NamespaceList{}
+
+	// This List will be from the cache
+	if err := r.Client.List(ctx, &namespaces); err != nil {
+		log.Error(err, "Unable to list namespaces from the cache")
+
+		return ctrl.Result{}, err
+	}
+
+	for name, oldSelection := range oldSelections {
+		newNamespaces, err := filter(namespaces, oldSelection.target)
+		if err != nil {
+			log.Error(err, "Unable to filter namespaces for policy", "name", name)
+
+			r.Update(name, NamespaceSelection{
+				target:     oldSelection.target,
+				namespaces: newNamespaces,
+				hasUpdate:  oldSelection.err != nil, // it has an update if the error state changed
+				err:        err,
+			})
+		}
+
+		if !reflect.DeepEqual(newNamespaces, oldSelection.namespaces) {
+			log.V(2).Info("Updating selection from Reconcile", "policy", name, "selection", newNamespaces)
+
+			r.Update(name, NamespaceSelection{
+				target:     oldSelection.target,
+				namespaces: newNamespaces,
+				hasUpdate:  true,
+				err:        nil,
+			})
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NamespaceSelectorReconciler) Get(name string, t policyv1.Target) ([]string, error) {
+	log := logf.Log.WithValues("Reconciler", "NamespaceSelector")
+
+	r.lock.Lock()
+
+	// If found, and target has not been changed
+	if selection, found := r.selections[name]; found && selection.target.String() == t.String() {
+		selection.hasUpdate = false
+		r.selections[name] = selection
+
+		r.lock.Unlock()
+
+		return selection.namespaces, selection.err
+	}
+
+	// unlock for now, the list filtering could take a non-trivial amount of time
+	r.lock.Unlock()
+
+	// New, or the target has changed.
+	namespaces := corev1.NamespaceList{}
+
+	// This List will be from the cache
+	if err := r.Client.List(context.TODO(), &namespaces); err != nil {
```

Comment:
You can use a label selector in the query when a label selector is used and not the include/exclude approach.

Reply:
I don't know which happens, but I believe that setting a selector here would either:
- do a list against the API server, which is then watched and cached, or
- just use the cache, and "locally" handle the selector

I definitely don't want the extra API server list or watch, that seems unnecessary. And I don't think there would be much of a memory consumption difference if the cache somehow does the selector, or if this controller does it in the `filter` function.

Or am I missing something?

Reply:
Discussed about this outside of GitHub, but the summary is:
* The cache uses the existing watch and performs the label selector matching locally.
* This local match occurs before the cache deep copies the objects returned from the `List` method so memory consumption is decreased as opposed to filtering after the `List` method.